ethpm_types/abi.py

```diff
@@ -18,69 +22,150 @@ def canonical_type(self) -> str:
             # Recursively discover the canonical type
             return self.type.canonical_type
 
+    @property
+    def signature(self) -> str:
+        if self.name:
+            return f"{self.canonical_type} {self.name}"
+        else:
+            return self.canonical_type
+
+
+class EventABIType(BaseModel):
+    name: Optional[str] = None  # NOTE: Tuples don't have names by default
+    indexed: bool = False  # Only event ABI types should have this field
+    type: Union[str, "ABIType"]
+    internalType: Optional[str] = None
+
+    @property
+    def canonical_type(self) -> str:
+        if isinstance(self.type, str):
+            return self.type
+
+        else:
+            # Recursively discover the canonical type
+            return self.type.canonical_type
```

Comment:
Would this make sense? And similar treatment for the ABI base class?

```suggestion
class ABIType(BaseModel):
    name: Optional[str] = None  # NOTE: Tuples don't have names by default
    type: Union[str, "ABIType"]
    internalType: Optional[str] = None

    @property
    def canonical_type(self) -> str:
        if isinstance(self.type, str):
            return self.type
        else:
            # Recursively discover the canonical type
            return self.type.canonical_type

    @property
    def signature(self) -> str:
        if self.name:
            return f"{self.canonical_type} {self.name}"
        else:
            return self.canonical_type


class EventABIType(ABIType):
    indexed: bool = False  # Only event ABI types should have this field

    @property
    def signature(self) -> str:
        sig = self.canonical_type
        # For events (handles both None and False conditions)
        if self.indexed:
            sig += " indexed"
        if self.name:
            sig += f" {self.name}"
        return sig

```

Reply:
Fair, I think this could work yeah